add-module-info failure when JAR is already modular

Invoking ModiTect's `add-module-info` goal as part of 2 separate build sessions causes an error

```
Error: /tmp/moditect-mcve/target/helloworld-1.0.0.jar is a modular JAR file that cannot be specified with the --generate-module-info option
```

jdeps fails because the JAR was made modular on the 1st build session.

Full MCVE available at https://github.com/aalmiray/moditect-mcve

```
$ gm build
Using maven at '/Users/aalmiray/.sdkman/candidates/maven/3.8.7/bin/mvn' to run buildFile '/tmp/moditect-mcve/pom.xml':
[INFO] Scanning for projects...
[INFO] 
[INFO] ------------------------< com.acme:helloworld >-------------------------
[INFO] Building helloworld 1.0.0
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ helloworld ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /tmp/moditect-mcve/src/main/resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.10.1:compile (default-compile) @ helloworld ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 1 source file to /tmp/moditect-mcve/target/classes
[INFO] 
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ helloworld ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /tmp/moditect-mcve/src/test/resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.10.1:testCompile (default-testCompile) @ helloworld ---
[INFO] No sources to compile
[INFO] 
[INFO] --- maven-surefire-plugin:2.12.4:test (default-test) @ helloworld ---
[INFO] No tests to run.
[INFO] 
[INFO] --- maven-jar-plugin:2.4:jar (default-jar) @ helloworld ---
[INFO] Building jar: /tmp/moditect-mcve/target/helloworld-1.0.0.jar
[INFO] 
[INFO] --- moditect-maven-plugin:1.0.0.RC2:add-module-info (add-module-infos) @ helloworld ---
writing to /tmp/moditect-mcve/target/moditect/helloworld/versions/9/module-info.java
[INFO] Created module descriptor at /tmp/moditect-mcve/target/moditect/generated-sources/com.acme.helloworld/module-info.java
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  1.689 s
[INFO] Finished at: 2023-03-01T20:02:29+01:00
[INFO] ------------------------------------------------------------------------

$ gm build
Using maven at '/Users/aalmiray/.sdkman/candidates/maven/3.8.7/bin/mvn' to run buildFile '/tmp/moditect-mcve/pom.xml':
[INFO] Scanning for projects...
[INFO] 
[INFO] ------------------------< com.acme:helloworld >-------------------------
[INFO] Building helloworld 1.0.0
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ helloworld ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /tmp/moditect-mcve/src/main/resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.10.1:compile (default-compile) @ helloworld ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ helloworld ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /tmp/moditect-mcve/src/test/resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.10.1:testCompile (default-testCompile) @ helloworld ---
[INFO] No sources to compile
[INFO] 
[INFO] --- maven-surefire-plugin:2.12.4:test (default-test) @ helloworld ---
[INFO] No tests to run.
[INFO] 
[INFO] --- maven-jar-plugin:2.4:jar (default-jar) @ helloworld ---
[INFO] 
[INFO] --- moditect-maven-plugin:1.0.0.RC2:add-module-info (add-module-infos) @ helloworld ---
Error: /tmp/moditect-mcve/target/helloworld-1.0.0.jar is a modular JAR file that cannot be specified with the --generate-module-info option
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  0.740 s
[INFO] Finished at: 2023-03-01T20:02:35+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.moditect:moditect-maven-plugin:1.0.0.RC2:add-module-info (add-module-infos) on project helloworld: Execution add-module-infos of goal org.moditect:moditect-maven-plugin:1.0.0.RC2:add-module-info failed: Invocation of jdeps failed: jdeps --generate-module-info /tmp/moditect-mcve/target/moditect --multi-release=9 /tmp/moditect-mcve/target/helloworld-1.0.0.jar -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
```

--- a/README.md
+++ b/README.md
@@ -211,6 +211,7 @@ the _add-module-info_ goal as follows:
             </goals>
             <configuration>
                 <jvmVersion>11</jvmVersion>
+                <failOnWarning>false</failOnWarning>
                 <module>
                     <moduleInfo>
                         <name>com.example</name>
@@ -241,7 +242,9 @@ The special value `base` (the default) can be used to add the descriptor to the 
 Putting the descriptor under `META-INF/versions` can help to increase compatibility with older libraries scanning class files that may fail when encountering the `module-info.class` file
 (as chances are lower that such tool will look for class files under `META-INF/versions/...`).
 
-The (optional) `exclusions` option may be used to filter out any `compile` or `runtime` dependencies that should not be used, as it might be the case when shading internal dependencies.
+The optional `failOnWarning` option prevents the build from failing when set to `false`. The default is to fail.
+
+The optional `exclusions` option may be used to filter out any `compile` or `runtime` dependencies that should not be used, as it might be the case when shading internal dependencies.
 
 The `jdepsExtraArgs` option can be used to specify a list of arguments passed to the _jdeps_ invocation for creating a "candidate descriptor".
 

--- a/maven-plugin/src/main/java/org/moditect/mavenplugin/add/AddModuleInfoMojo.java
+++ b/maven-plugin/src/main/java/org/moditect/mavenplugin/add/AddModuleInfoMojo.java
@@ -21,6 +21,7 @@ import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.StandardCopyOption;
 import java.util.Arrays;
+import java.util.Enumeration;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
@@ -28,6 +29,9 @@ import java.util.Map.Entry;
 import java.util.Objects;
 import java.util.Optional;
 import java.util.Set;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 
 import org.apache.maven.artifact.Artifact;
@@ -67,6 +71,10 @@ import org.moditect.model.GeneratedModuleInfo;
 @Mojo(name = "add-module-info", defaultPhase = LifecyclePhase.PROCESS_RESOURCES, requiresDependencyResolution = ResolutionScope.COMPILE_PLUS_RUNTIME)
 public class AddModuleInfoMojo extends AbstractMojo {
 
+    private static final String MODULE_INFO_CLASS = "module-info.class";
+
+    private static final Pattern VERSIONED_MODULE_INFO = Pattern.compile( "META-INF/versions/\\d+/" + MODULE_INFO_CLASS );
+
     @Component
     private RepositorySystem repoSystem;
 
@@ -102,6 +110,9 @@ public class AddModuleInfoMojo extends AbstractMojo {
 
     @Parameter(property = "moditect.skip", defaultValue = "false")
     private boolean skip;
+
+    @Parameter(property = "moditect.failOnWarning", defaultValue = "true")
+    private boolean failOnWarning;
 
     @Parameter
     private MainModuleConfiguration module;
@@ -153,6 +164,16 @@ public class AddModuleInfoMojo extends AbstractMojo {
         if ( modules != null ) {
             for ( ModuleConfiguration moduleConfiguration : modules ) {
                 Path inputFile = getInputFile( moduleConfiguration, artifactResolutionHelper );
+                if ( isModularJar( inputFile ) ) {
+                    String message = "File " + inputFile.getFileName() + " is already modular";
+                    if ( failOnWarning ) {
+                        throw new MojoExecutionException( message );
+                    } else {
+                        getLog().warn( message );
+                        continue;
+                    }
+                }
+
                 String moduleInfoSource = getModuleInfoSource( inputFile, moduleConfiguration, moduleInfoGenerator, assignedNamesByModule, modularizedJars );
 
                 AddModuleInfo addModuleInfo = new AddModuleInfo(
@@ -182,6 +203,16 @@ public class AddModuleInfoMojo extends AbstractMojo {
                 throw new MojoExecutionException( "Couldn't find file " + inputJar + ". Run this goal for the project's JAR only after the maven-jar-plugin." );
             }
 
+            if ( isModularJar( inputJar ) ) {
+                String message = "File " + inputJar.getFileName() + " is already modular";
+                if ( failOnWarning ) {
+                    throw new MojoExecutionException( message );
+                } else {
+                    getLog().warn( message );
+                    return;
+                }
+            }
+
             AddModuleInfo addModuleInfo = new AddModuleInfo(
                     getModuleInfoSource( inputJar, module, moduleInfoGenerator, assignedNamesByModule, modularizedJars ),
                     module.getMainClass(),
@@ -200,6 +231,27 @@ public class AddModuleInfoMojo extends AbstractMojo {
                 throw new RuntimeException( "Couldn't replace " + inputJar + " with modularized version", e );
             }
         }
+    }
+
+    /**
+     * Determines if the given JAR is already modular that is,
+     * it contains either a "module-info.class" entry or at least
+     * one "/META-INF/version/\\d+/module-info.class" entry.
+     */
+    private boolean isModularJar(Path jarPath) {
+        try( JarFile jarFile = new JarFile( jarPath.toFile() ) ) {
+            Enumeration<JarEntry> entries = jarFile.entries();
+            while ( entries.hasMoreElements() ) {
+                JarEntry entry = entries.nextElement();
+                String entryName = entry.getName();
+                if ( entryName.equals( MODULE_INFO_CLASS ) || VERSIONED_MODULE_INFO.matcher( entryName ).matches() ) {
+                    return true;
+                }
+            }
+        } catch ( IOException ioe ) {
+            throw new IllegalStateException( "Unexpected error when reading " + jarPath.getFileName() );
+        }
+        return false;
     }
 
     /**

